File: ios-app/Services/CurrencyFormatter.swift
import Foundation

/// Centralized currency formatting — mirrors the Android CurrencyFormatter and web's lib/currency.ts.
///
/// Usage:
/// ```
/// CurrencyFormatter.shared.formatSimple(1234.50)         // "KES 1,234.50"
/// CurrencyFormatter.shared.formatSimple(1234.50, "USD")  // "USD 1,234.50"
/// CurrencyFormatter.shared.format(1234.50, "KES")        // "KSh1,234.50" (locale-aware)
/// ```
final class CurrencyFormatter {
    static let shared = CurrencyFormatter()
    
    /// Default currency code — updated by WorkspaceManager when the active workspace changes.
    /// Falls back to "KES" until a workspace is loaded from the API.
    var defaultCurrencyCode: String = "KES"
    
    private let localeMap: [String: Locale] = [
        "KES": Locale(identifier: "en_KE"),
        "USD": Locale(identifier: "en_US"),
        "EUR": Locale(identifier: "de_DE"),
        "GBP": Locale(identifier: "en_GB"),
        "JPY": Locale(identifier: "ja_JP"),
        "AUD": Locale(identifier: "en_AU"),
        "CAD": Locale(identifier: "en_CA"),
        "CHF": Locale(identifier: "de_CH"),
        "CNY": Locale(identifier: "zh_CN"),
        "INR": Locale(identifier: "en_IN"),
        "ZAR": Locale(identifier: "en_ZA"),
        "NGN": Locale(identifier: "en_NG"),
        "TZS": Locale(identifier: "en_TZ"),
        "UGX": Locale(identifier: "en_UG"),
    ]
    
    private var formatters: [String: NumberFormatter] = [:]
    
    /// Locale-aware format: "KSh1,234.50" or "$1,234.50" depending on locale.
    func format(_ amount: Double, _ currencyCode: String? = nil) -> String {
        let code = currencyCode ?? defaultCurrencyCode
        if formatters[code] == nil {
            let nf = NumberFormatter()
            nf.numberStyle = .currency
            nf.locale = localeMap[code] ?? Locale(identifier: "en_US")
            nf.currencyCode = code
            nf.minimumFractionDigits = 2
            nf.maximumFractionDigits = 2
            formatters[code] = nf
        }
        return formatters[code]!.string(from: NSNumber(value: amount)) ?? "\(code) \(amount)"
    }
    
    /// Simple format: "KES 1,234.50" — always uses the code prefix, consistent cross-platform.
    func formatSimple(_ amount: Double, _ currencyCode: String? = nil) -> String {
        let code = currencyCode ?? defaultCurrencyCode
        let nf = NumberFormatter()
        nf.numberStyle = .decimal
        nf.minimumFractionDigits = 2
        nf.maximumFractionDigits = 2
        nf.groupingSeparator = ","
        nf.decimalSeparator = "."
        let formatted = nf.string(from: NSNumber(value: amount)) ?? String(format: "%.2f", amount)
        return "\(code) \(formatted)"
    }
}

File: android-app/app/src/main/java/com/mafutapass/app/util/CurrencyFormatter.kt
package com.mafutapass.app.util

import java.text.NumberFormat
import java.util.Currency
import java.util.Locale

/**
 * Format an amount with a currency code.
 * Uses Intl NumberFormat under the hood — consistent with the web app's lib/currency.ts.
 *
 * Usage: formatCurrency(1234.50, "KES") → "KES 1,234.50"
 */
object CurrencyFormatter {

    /**
     * Default currency code used when callers omit the currencyCode parameter.
     * Updated automatically by WorkspaceRepository when the active workspace changes.
     * Defaults to "KES" until the workspace is fetched from the API.
     */
    @Volatile
    var defaultCurrencyCode: String = "KES"

    private val formatters = mutableMapOf<String, NumberFormat>()

    private val localeMap = mapOf(
        "KES" to Locale("en", "KE"),
        "USD" to Locale.US,
        "EUR" to Locale.GERMANY,
        "GBP" to Locale.UK,
        "JPY" to Locale.JAPAN,
        "AUD" to Locale("en", "AU"),
        "CAD" to Locale.CANADA,
        "CHF" to Locale("de", "CH"),
        "CNY" to Locale.CHINA,
        "INR" to Locale("en", "IN"),
        "ZAR" to Locale("en", "ZA"),
        "NGN" to Locale("en", "NG"),
        "TZS" to Locale("en", "TZ"),
        "UGX" to Locale("en", "UG"),
    )

    fun format(amount: Double, currencyCode: String = defaultCurrencyCode): String {
        val formatter = formatters.getOrPut(currencyCode) {
            val locale = localeMap[currencyCode] ?: Locale.US
            NumberFormat.getCurrencyInstance(locale).apply {
                try {
                    currency = Currency.getInstance(currencyCode)
                } catch (_: Exception) {
                    // Fall back to locale default
                }
                minimumFractionDigits = 2
                maximumFractionDigits = 2
            }
        }
        return synchronized(formatter) { formatter.format(amount) }
    }

    /**
     * Simple format: "KES 1,234.50" — always uses the code prefix.
     * Useful when the platform NumberFormat produces symbols the user won't recognize.
     */
    fun formatSimple(amount: Double, currencyCode: String = defaultCurrencyCode): String {
        return "$currencyCode ${String.format("%,.2f", amount)}"
    }
}
